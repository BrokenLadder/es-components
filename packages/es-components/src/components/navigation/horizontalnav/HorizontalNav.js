/* eslint react/no-unused-prop-types: 0 */
/* ^^^ This is because defaultSelected needs to be defined to be documented */
/* The behavior of defaultSelected is implemented in the Uncontrollable library */

import React, { Children, cloneElement } from 'react';
import PropTypes from 'prop-types';
import styled, { withTheme } from 'styled-components';
import { noop } from 'lodash';
import uncontrollable from 'uncontrollable';

import { HorizontalNavItem } from './HorizontalNavItem';

const NavStyled = styled.nav`
  background-color: ${props => props.theme.colors.white};
  ul {
    display: flex;
    list-style: none;
    margin: 0;
    padding: 0;
  }
`;

export function HorizontalNav(props) {
  const { useAltStyle, children, onItemSelected, selected } = props;

  const onNavClick = (navId = null) => {
    onItemSelected(navId);
  };

  return (
    <NavStyled className="es-horizontalnav">
      <ul>
        {Children.toArray(children).map(child => {
          if (child !== null && child.type === HorizontalNavItem) {
            return cloneElement(child, {
              useAltStyle,
              highlightedId: selected,
              onNavClick
            });
          }
          return child;
        })}
      </ul>
    </NavStyled>
  );
}

HorizontalNav.propTypes = {
  /** Use the alternate nav style */
  useAltStyle: PropTypes.bool,
  children: PropTypes.node,
  /** Set the selected nav item by id, controlled mode */
  selected: PropTypes.string,
  /** Use to set a default nav, uncontrolled mode */
  defaultSelected: PropTypes.string,
  /** Function called when a nav item is clicked */
  onItemSelected: PropTypes.func
};

HorizontalNav.defaultProps = {
  useAltStyle: false,
  onItemSelected: noop,
  children: undefined,
  selected: undefined,
  defaultSelected: undefined
};

const UncontrolledHorizontalNav = uncontrollable(HorizontalNav, {
  selected: 'onItemSelected'
});

UncontrolledHorizontalNav.Item = HorizontalNavItem;

export default withTheme(UncontrolledHorizontalNav);
