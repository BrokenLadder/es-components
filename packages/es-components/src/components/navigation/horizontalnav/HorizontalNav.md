(Uncontrolled) Use `defaultSelected` to set an initial selected item.
```
<div>
  <div>
    <h3>Standard</h3>
    <HorizontalNav defaultSelected="home">
      <HorizontalNav.Item id="home" targetUrl="#home">Home</HorizontalNav.Item>
      <HorizontalNav.Item id="cart">Cart</HorizontalNav.Item>
      <HorizontalNav.Item id="help">Help</HorizontalNav.Item>
      <HorizontalNav.Item id="info">About</HorizontalNav.Item>
      <HorizontalNav.Item id="disabled" isDisabled>Disabled</HorizontalNav.Item>
    </HorizontalNav>
  </div>

  <div>
    <h3>Alternate Style</h3>
    <HorizontalNav useAltStyle defaultSelected="snapshot">
      <HorizontalNav.Item id="snapshot">Plan Snapshot</HorizontalNav.Item>
      <HorizontalNav.Item id="medical">Medical Benefits</HorizontalNav.Item>
      <HorizontalNav.Item id="drug">Drug Benefits</HorizontalNav.Item>
      <HorizontalNav.Item id="docs">Plan Documents</HorizontalNav.Item>
      <HorizontalNav.Item id="disabled" isDisabled>Disabled</HorizontalNav.Item>
    </HorizontalNav>
  </div>
</div>
```

(Controlled) Use `onItemSelected` to control the `selected` value.
```
class NavExample extends React.Component {
  constructor() {
    this.state = { selected: "home" };
    this.onItemSelected = this.onItemSelected.bind(this);
  }

  onItemSelected(value) {
    this.setState({ selected: value });
    alert(`${value} selected!`);
  };

  render() {
    return (
      <div>
        <HorizontalNav selected={ this.state.selected } onItemSelected={ this.onItemSelected }>
          <HorizontalNav.Item id="home"><Icon name="home" /> Home</HorizontalNav.Item>
          <HorizontalNav.Item id="cart"><Icon name="shopping-cart" /> Cart</HorizontalNav.Item>
          <HorizontalNav.Item id="disabled" isDisabled><Icon name="no-symbol" /> Disabled</HorizontalNav.Item>
        </HorizontalNav>
      </div>
    )
  }
}
<NavExample />
```

Set `onClick` on an individual `HorizontalNav.Item`.
```
<div>
  <HorizontalNav useAltStyle>
    <HorizontalNav.Item id="home" onClick={ () => (alert('Time to go Home!'))}>Home</HorizontalNav.Item>
    <HorizontalNav.Item id="cart" onClick={ () => (alert('You clicked the Cart!'))}>Cart</HorizontalNav.Item>
    <HorizontalNav.Item id="disabled" isDisabled>Disabled</HorizontalNav.Item>
  </HorizontalNav>
</div>
```

Assign a `targetUrl` to set the nav link href location. Use `isExternalLink` to open the link in a new browser window.
```
<div>
  <HorizontalNav>
    <HorizontalNav.Item id="home" targetUrl="/">Home</HorizontalNav.Item>
    <HorizontalNav.Item id="medicare" targetUrl="https://medicare.oneexchange.com" isExternalLink>Medicare Site</HorizontalNav.Item>
    <HorizontalNav.Item id="sidenav" targetUrl="#sidenav">HorizontalNav Section</HorizontalNav.Item>
  </HorizontalNav>
</div>
```
