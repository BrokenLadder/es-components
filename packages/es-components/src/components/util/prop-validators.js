import isParseableDate from './isParseableDate';

export function isPropValidDate(propsList, propName, componentName) {
  const date = propsList[propName];
  if (date) {
    if (!isParseableDate(date)) {
      return new Error(`The pre-selected date is not valid. Please enter a parseable date.
        See http://en.wikipedia.org/wiki/ISO_8601 for valid date formats.
      `);
    }
    return null;
  }
  return null;
}
