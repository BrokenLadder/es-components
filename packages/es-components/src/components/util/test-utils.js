import React from 'react';
import { ThemeProvider } from 'styled-components';
import viaTheme from 'es-components-via-theme';
import { render } from 'react-testing-library';

export function renderWithTheme(component) {
  return render(
    <ThemeProvider theme={viaTheme}>
      {React.cloneElement(component)}
    </ThemeProvider>
  )
}