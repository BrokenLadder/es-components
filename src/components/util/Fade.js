import React from 'react';
import PropTypes from 'prop-types';
import Transition from 'react-overlays/lib/Transition';
import { injectGlobal } from 'styled-components';

/* eslint-disable no-unused-expressions */
injectGlobal`
  .default-transition-out {
    opacity: 0;
    transition: opacity 300ms linear;
  }

  .default-transition-in {
    opacity: 1;
  }
`;
/* eslint-enable no-unused-expressions */

const Fade = props => {
  const {
    in: transitionIn,
    timeout,
    transitionClassIn,
    transitionClassOut,
    ...otherProps
  } = props;

  return (
    <Transition
      className={transitionClassOut}
      enteredClassName={transitionClassIn}
      enteringClassName={transitionClassIn}
      in={transitionIn}
      unmountOnExit
      timeout={timeout}
      transitionAppear
      {...otherProps}
    />
  );
};

Fade.propTypes = {
  children: PropTypes.any.isRequired,
  in: PropTypes.bool,
  transitionClassIn: PropTypes.string,
  transitionClassOut: PropTypes.string,
  timeout: PropTypes.number
};

Fade.defaultProps = {
  transitionClassIn: 'default-transition-in',
  transitionClassOut: 'default-transition-out',
  timeout: 5000
};

export default Fade;
