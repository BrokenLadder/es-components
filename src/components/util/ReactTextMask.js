import React from 'react';
import ReactTextMask from 'react-text-mask';

// the component was using onInput which made testing more difficult.
// switched to onChange and put in a pull request so hopefully we can
// get rid of this soon.
ReactTextMask.prototype.render = function render() {
  const props = { ...this.props };

  delete props.mask;
  delete props.guide;
  delete props.pipe;
  delete props.placeholderChar;
  delete props.keepCharPositions;
  delete props.onChange;
  delete props.value;
  delete props.showMask;

  return (
    <input
      {...props}
      onChange={event => this.onChange(event)}
      defaultValue={this.props.value}
      ref={inputElement => (this.inputElement = inputElement)}
    />
  );
};

export default ReactTextMask;
