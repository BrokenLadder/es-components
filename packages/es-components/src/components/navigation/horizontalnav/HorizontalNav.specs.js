/* eslint-env jest */

import React from 'react';
import { renderWithTheme } from '../../util/test-utils';

import HorizontalNav from './HorizontalNav';

let instanceToRender;
const onItemSelected = jest.fn();
const onClick = jest.fn();

beforeEach(() => {
  onItemSelected.mockClear();
  onClick.mockClear();

  instanceToRender = (
    <HorizontalNav onItemSelected={navId => onItemSelected(navId)}>
      <HorizontalNav.Item
        id="home"
        className="home"
        onClick={navId => onClick(navId)}
        targetUrl="/home"
      >
        Home
      </HorizontalNav.Item>
      <HorizontalNav.Item id="cart" className="cart">
        Cart
      </HorizontalNav.Item>
      <HorizontalNav.Item
        id="external"
        className="external"
        targetUrl="http://www.google.com"
        isExternalLink
      >
        External Link
      </HorizontalNav.Item>
      <HorizontalNav.Item
        id="info"
        className="info"
        onClick={() => onClick()}
        isDisabled
      >
        Disabled
      </HorizontalNav.Item>
    </HorizontalNav>
  );
});

it('renders as expected', () => {
  const { container } = renderWithTheme(instanceToRender);
  expect(container).toMatchSnapshot();
});

it('executes onItemSelected with the id of the nav item clicked', () => {
  const { getByText } = renderWithTheme(instanceToRender);
  getByText('Cart').click();
  expect(onItemSelected).toBeCalledWith('cart');
});

it('executes onClick when nav item clicked', () => {
  const { getByText } = renderWithTheme(instanceToRender);
  getByText('Home').click();
  expect(onClick).toBeCalledWith('home');
});

it('disabled item prevents onclick functions', () => {
  const { getByText } = renderWithTheme(instanceToRender);
  getByText('Disabled').click();
  expect(onClick).not.toBeCalled();
  expect(onItemSelected).not.toBeCalled();
});
