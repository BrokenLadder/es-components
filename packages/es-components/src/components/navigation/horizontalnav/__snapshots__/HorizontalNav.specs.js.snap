// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <nav
    class="es-horizontalnav sc-bxivhb cMujCD"
  >
    <ul>
      <li
        class="es-horizontalnav__navitem"
      >
        <a
          class="home sc-bdVaJa cisBhB"
          href="/home"
          target="_self"
        >
          <span
            class="sc-htpNat kbKuoc"
          >
            Home
          </span>
        </a>
      </li>
      <li
        class="es-horizontalnav__navitem"
      >
        <a
          class="cart sc-bdVaJa cisBhB"
          href="#/"
          target="_self"
        >
          <span
            class="sc-htpNat kbKuoc"
          >
            Cart
          </span>
        </a>
      </li>
      <li
        class="es-horizontalnav__navitem"
      >
        <a
          class="external sc-bdVaJa cisBhB"
          href="http://www.google.com"
          target="_blank"
        >
          <span
            class="sc-htpNat kbKuoc"
          >
            External Link
          </span>
        </a>
      </li>
      <li
        class="es-horizontalnav__navitem"
      >
        <a
          class="info sc-bdVaJa bQVDYh"
          href="#/"
          target="_self"
        >
          <span
            class="sc-htpNat kbKuoc"
          >
            Disabled
          </span>
        </a>
      </li>
    </ul>
  </nav>
</div>
`;
